src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -357,6 +358,159 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncode(char* s, int length)
+        {
+            int numUnsafe = 0;
+            int numSpaces = 0;
+            for (int i = 0; i < length; i++)
+            {
+                char c = s[i];
+                if (c == ' ')
+                {
+                    numSpaces++;
+                }
+                else if (!IsUrlSafeChar(s[i]))
+                {
+                    numUnsafe += GetByteCount(s, i, length);
```

Comment:
> I'm under the impression they're not taking mscorlib contributions (at least until the release of 1.0)

We can always discuss changes, and then decide whether and when it makes sense to take them.

I have written the current UTF8Encoding in mscorlib myself together with a few other people. It does number of perf tricks like processing multiple characters at a time - it should beat your impl on large strings. I believe that most of the overhead left in it is imposed by the surface area (abstract base class, customizable encoding fallbacks) - you can beat it on small strings because of this.